src/v/raft/mux_state_machine.h

```diff
@@ -214,31 +209,57 @@ ss::future<std::error_code> mux_state_machine<T...>::replicate_and_wait(
-    using ret_t = std::error_code;
-    return ss::with_gate(
-      _gate, [this, b = std::move(b), timeout, &as]() mutable {
-          return _mutex.get_units().then([this, b = std::move(b), timeout, &as](
-                                           ss::semaphore_units<> u) mutable {
-              return replicate(std::move(b))
-                .then([this, u = std::move(u), timeout, &as](
-                        result<raft::replicate_result> r) {
-                    if (!r) {
-                        return ss::make_ready_future<ret_t>(r.error());
-                    }
-
-                    auto last_offset = r.value().last_offset;
-
-                    auto [it, insterted] = _promises.emplace(
-                      last_offset, expiring_promise<ret_t>{});
-                    vassert(
-                      insterted,
-                      "Prosmise for offset {} already registered",
-                      last_offset);
-                    return it->second.get_future_with_timeout(
-                      timeout, [] { return errc::timeout; }, as);
-                });
-          });
+    if (_gate.is_closed()) {
+        return ss::make_ready_future<std::error_code>(errc::shutting_down);
+    }
+
+    auto u = get_units();
+    auto promise = std::make_unique<promise_t>();
+    auto f = promise->get_future_with_timeout(
+      timeout, [] { return make_error_code(errc::timeout); }, as);
+
+    ssx::spawn_with_gate(
+      _gate,
+      [this,
+       b = std::move(b),
+       u = std::move(u),
+       promise = std::move(promise)]() mutable {
+          return replicate(std::move(b))
+            .then([this, u = std::move(u), promise = std::move(promise)](
+                    result<raft::replicate_result> r) mutable {
+                if (!r) {
+                    promise->set_value(r.error());
+                    return ss::now();
+                }
+
+                auto last_offset = r.value().last_offset;
+
+                return _result
+                  .wait([this, last_offset] {
+                      return _last_applied >= last_offset;
+                  })
+                  .then([this,
+                         last_offset,
+                         u = std::move(u),
+                         promise = std::move(promise)]() mutable {
+                      u.release();
+                      auto it = _results.find(last_offset);
+                      vassert(
+                        it != _results.end(),
+                        "last applied offset {} is greater than "
+                        "returned replicate result {}. this must "
+                        "imply existence of a result in results map",
+                        _last_applied,
+                        last_offset);
+                      auto res = it->second;
+                      if (_pending == 0) {
+                          _results.clear();
+                      }
+                      promise->set_value(res);
```

Comment:
Is there a reason we can't do `_results.erase(it)` here?

Reply:
actually we can